test/subscription.js

```diff
@@ -1097,3 +1097,309 @@ test('subscription works properly if onConnect is not defined and connectionInit
+      payload: {
+        query: `
+            subscription {
+              notificationAdded(contains: "Hello") {
+                id
+                message
+              }
+            }
+          `
+      }
+    }))
+
+    client.write(JSON.stringify({
+      id: 2,
+      type: 'start',
+      payload: {
+        query: `
+            subscription {
+              notificationAdded {
+                id
+                message
+              }
+            }
+          `
+      }
+    }))
+
+    client.write(JSON.stringify({
+      id: 2,
+      type: 'stop'
+    }))
+
+    client.on('data', chunk => {
+      const data = JSON.parse(chunk)
+
+      if (data.id === 1 && data.type === 'data') {
+        t.true(!data.payload.data.notificationAdded.message.includes('filtered'))
+        client.end()
+      } else if (data.id === 2 && data.type === 'complete') {
+        app.inject({
+          method: 'POST',
+          url: '/graphql',
+          body: {
+            query: `
+              mutation {
+                addNotification(message: "filtered should not pass") {
+                  id
+                }
+              }
+            `
+          }
+        }, () => { t.pass() })
+        app.inject({
+          method: 'POST',
+          url: '/graphql',
+          body: {
+            query: `
+              mutation {
+                addNotification(message: "Hello World") {
+                  id
+                }
+              }
+            `
+          }
+        }, () => { t.pass() })
+      }
+    })
+  })
+})
+
+test('subscription handles `withFilter` if filter throws', t => {
+  t.plan(4)
+  const app = Fastify()
+  t.tearDown(() => app.close())
+
+  const { withFilter } = GQL
+
+  let idCount = 0
+  const notifications = []
+
+  const schema = `
+    type Notification {
+      id: ID!
+      message: String
+    }
+
+    type Query {
+      notifications: [Notification]
+    }
+
+    type Mutation {
+      addNotification(message: String!): Notification!
+    }
+
+    type Subscription {
+      notificationAdded(contains: String): Notification
+    }
+  `
+
+  const resolvers = {
+    Query: {
+      notifications: () => notifications
+    },
+    Mutation: {
+      addNotification: async (_, { message }, { pubsub }) => {
+        const id = idCount++
+        const notification = {
+          id,
+          message
+        }
+        notifications.push(notification)
+        await pubsub.publish({
+          topic: 'NOTIFICATION_ADDED',
+          payload: {
+            notificationAdded: notification
+          }
+        })
+      }
+    },
+    Subscription: {
+      notificationAdded: {
+        subscribe: withFilter(
+          (_, __, { pubsub }) => pubsub.subscribe('NOTIFICATION_ADDED'),
+          (payload, { contains }) => {
+            if (contains && !payload.notificationAdded.message.includes(contains)) {
+              throw new Error('fail')
+            }
+            return true
```

Comment:
Can you add a test where this returns a Promise? I'm not really sure this should be `async`, it might just be a synchronous implementation instead.

Reply:
If we go for a synchronous implementation, I think this is achievable with a stream Transform. But IMO this should be async if some async stuff is required to filter the notification (apollo's implementation allows async filter). I'll add an async test to see if it works as expected.

Reply:
Unfortunately a stream transform is slower than an async iterator, so better go with the async iterator anyway.

Reply:
Let's add a test for the async version

Reply:
> Unfortunately a stream transform is slower than an async iterator, so better go with the async iterator anyway.

Ok, good to know ^^